### PR TITLE
add search module bar into landscape screen

### LINKED ISSUE #11102: Add a search module bar to the landscape screen
~~### Prerequisite~~

~~- [ ] Add a button to hide the properties form component (needed on small device screens)~~
~~- [ ] Memorize in local storage the hide/show preference options selected by the user from the minimap/presetConfiguration/propertiesForm components. This will avoid the annoying situation of having to hide or show components every time they press F5 or access JHipster Lite. `ComponentsVisibilityRepository`-> `LocalStorageComponentsVisibilityRepository`~~
~~- [ ] Use an icon on the minimap instead of the large `OPEN MINIMAP` button to optimize screen space~~

### TODO (edited: 20241012)

- [x] Implement a search functionality in the modules bar that identifies the module/feature by name as each letter is typed
- [x] Highlight the first module/feature that matches the typed letters in the search modules bar
- [x] Scroll the screen to the highlighted module if it is not visible within the current viewport
- [x] Reset the screen position to the initial landscape layout when the search bar is cleared
- [x] Remove the hilight when the search bar is cleared.
- [x] Hide the search modules bar on devices with screens smaller than or equal to 800px, as it becomes impractical to use without the option to hide other components

### Demo of the prototype to show the viability of this feature

The recorded video missed some frames and is lagging, but it's super smooth to use. Please ignore the search button, as it wasn't needed.

https://github.com/user-attachments/assets/d3ab75c1-0138-4628-a67e-d769ba6be938

### PR DESCRIPTION
**Pull request by renanfranca**
- Fix #11102 

A demo video showing how it works (in Full HD and smoothly captured with my Graphics card):

https://github.com/user-attachments/assets/9822ee47-2d0b-4c35-92aa-db02a321839a

I did my best here, and I hope it is a great enhancement to the landscape screen experience. The search module bar does not show on small screens:

image

### CODE REVIEW

File: src/main/webapp/app/module/primary/landscape/Landscape.component.ts [hunk 1]
```diff
@@ -592,6 +598,44 @@ export default defineComponent({
...
+      highlightedModule.value = foundModule ? new ModuleSlug(foundModule) : null;
+
+      if (foundModule) {
+        const moduleElement = landscapeElements.value.get(foundModule)!;
+        nextTick().then(() => scrollToHighlightedModule(moduleElement));
+      }
+    };
+
+    const findModule = (query: string): string | null => {
+      return [...landscapeElements.value.keys()].find(key => key.toLowerCase().includes(query.toLowerCase())) || null;
```
Comment (qmonmert) [inline src/main/webapp/app/module/primary/landscape/Landscape.component.ts:622]:
```suggestion
      return [...landscapeElements.value.keys()].find(key => key.toLowerCase().includes(query.toLowerCase())) ?? null;
```
image

File: src/main/webapp/app/module/primary/landscape/Landscape.component.ts [hunk 2]
```diff
@@ -87,6 +87,8 @@ export default defineComponent({
     const selectedPreset = ref<Preset | null>(null);
     const selectedPresetName = computed(() => selectedPreset.value?.name ?? '');
 
+    const highlightedModule = ref<ModuleSlug | null>(null);
```
Comment (DamnClin) [inline src/main/webapp/app/module/primary/landscape/Landscape.component.ts:90]: I think this is an ref(Optional<ModuleSlug>)

File: src/main/webapp/app/module/primary/landscape/Landscape.component.ts [hunk 3]
```diff
@@ -384,6 +387,13 @@ export default defineComponent({
       return ' ' + modeClass();
     };
 
+    const searchHighlightClass = (module: LandscapeElementId): string => {
+      if (highlightedModule.value && module.get() === highlightedModule.value.get()) {
```
Comment (DamnClin) [inline src/main/webapp/app/module/primary/landscape/Landscape.component.ts:391]: with if beeign an Optional: highlightedModule.value.filter(highligted => highligted.get() === module.get()).map(() => ' -search-highlighted').orElse('')

File: src/main/webapp/app/module/primary/landscape/Landscape.component.ts [hunk 4]
```diff
@@ -592,6 +598,44 @@ export default defineComponent({
...
+      const foundModule = findModule(query);
+      highlightedModule.value = foundModule ? new ModuleSlug(foundModule) : null;
+
+      if (foundModule) {
+        const moduleElement = landscapeElements.value.get(foundModule)!;
+        nextTick().then(() => scrollToHighlightedModule(moduleElement));
+      }
+    };
+
+    const findModule = (query: string): string | null => {
```
Comment (DamnClin) [inline src/main/webapp/app/module/primary/landscape/Landscape.component.ts:621]: Easier to read with an Optional

**Comment by renanfranca** [comment]
Thank you, @qmonmert and @DamnClin, for the review! I made the changes, please, let me know if everything looks good now. 👍

PS: The use of Optional made the code much more elegant. 😄✌

**Review by qmonmert — APPROVED** [review]

**Pull request merged**